TestLeasingPutGetDeleteConcurrent

Was running `clientv3/integration` manually in my linux VM:

```
=== RUN   TestLeasingPutGetDeleteConcurrent
2017-08-23 22:36:49.064852 I | integration: launching 3852279067361483651 (unix://localhost:38522790673614836510)
2017-08-23 22:36:49.068746 I | etcdserver: name = 3852279067361483651
2017-08-23 22:36:49.068799 I | etcdserver: data dir = /tmp/etcd625417580
2017-08-23 22:36:49.068842 I | etcdserver: member dir = /tmp/etcd625417580/member
2017-08-23 22:36:49.068883 I | etcdserver: heartbeat = 10ms
2017-08-23 22:36:49.068920 I | etcdserver: election = 100ms
2017-08-23 22:36:49.068955 I | etcdserver: snapshot count = 0
2017-08-23 22:36:49.068999 I | etcdserver: advertise client URLs = unix://127.0.0.1:2128024691
2017-08-23 22:36:49.069067 I | etcdserver: initial advertise peer URLs = unix://127.0.0.1:2127924691
2017-08-23 22:36:49.069120 I | etcdserver: initial cluster = 3852279067361483651=unix://127.0.0.1:2127924691
2017-08-23 22:36:49.073813 I | etcdserver: starting member 40014c3fd3be6d85 in cluster 8b6479478601236b
2017-08-23 22:36:49.073900 I | raft: 40014c3fd3be6d85 became follower at term 0
2017-08-23 22:36:49.073956 I | raft: newRaft 40014c3fd3be6d85 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2017-08-23 22:36:49.074002 I | raft: 40014c3fd3be6d85 became follower at term 1
2017-08-23 22:36:49.080520 W | auth: simple token is not cryptographically signed
2017-08-23 22:36:49.083509 I | etcdserver: set snapshot count to default 100000
2017-08-23 22:36:49.083570 I | etcdserver: starting server... [version: 3.2.0+git, cluster version: to_be_decided]
2017-08-23 22:36:49.084574 I | etcdserver/membership: added member 40014c3fd3be6d85 [unix://127.0.0.1:2127924691] to cluster 8b6479478601236b
2017-08-23 22:36:49.086923 I | integration: launched 3852279067361483651 (unix://localhost:38522790673614836510)
2017-08-23 22:36:49.164511 I | raft: 40014c3fd3be6d85 is starting a new election at term 1
2017-08-23 22:36:49.164601 I | raft: 40014c3fd3be6d85 became candidate at term 2
2017-08-23 22:36:49.164655 I | raft: 40014c3fd3be6d85 received MsgVoteResp from 40014c3fd3be6d85 at term 2
2017-08-23 22:36:49.164711 I | raft: 40014c3fd3be6d85 became leader at term 2
2017-08-23 22:36:49.164763 I | raft: raft.node: 40014c3fd3be6d85 elected leader 40014c3fd3be6d85 at term 2
2017-08-23 22:36:49.165572 I | etcdserver: setting up the initial cluster version to 3.2
2017-08-23 22:36:49.165713 I | etcdserver: published {Name:3852279067361483651 ClientURLs:[unix://127.0.0.1:2128024691]} to cluster 8b6479478601236b
2017-08-23 22:36:49.166247 N | etcdserver/membership: set the initial cluster version to 3.2
==================
WARNING: DATA RACE
Read at 0x00c422388ea0 by goroutine 55:
  runtime.chansend()
      /usr/local/go/src/runtime/chan.go:128 +0x0
  github.com/coreos/etcd/mvcc.(*watcher).send()
      /home/gyuho/go/src/github.com/coreos/etcd/mvcc/watchable_store.go:518 +0x4ac
  github.com/coreos/etcd/mvcc.(*watchableStore).notify()
      /home/gyuho/go/src/github.com/coreos/etcd/mvcc/watchable_store.go:429 +0x22c
  github.com/coreos/etcd/mvcc.(*watchableStoreTxnWrite).End()
      /home/gyuho/go/src/github.com/coreos/etcd/mvcc/watchable_store_txn.go:43 +0x384
  github.com/coreos/etcd/etcdserver.(*applierV3backend).Txn()
      /home/gyuho/go/src/github.com/coreos/etcd/etcdserver/apply.go:361 +0x360
  github.com/coreos/etcd/etcdserver.(*quotaApplierV3).Txn()
      /home/gyuho/go/src/github.com/coreos/etcd/etcdserver/apply.go:784 +0xaa
  github.com/coreos/etcd/etcdserver.(*authApplierV3).Txn()
      /home/gyuho/go/src/github.com/coreos/etcd/etcdserver/apply_auth.go:169 +0xed
  github.com/coreos/etcd/etcdserver.(*applierV3backend).Apply()
      /home/gyuho/go/src/github.com/coreos/etcd/etcdserver/apply.go:119 +0x624
  github.com/coreos/etcd/etcdserver.(*quotaApplierV3).Apply()
      <autogenerated>:1 +0x61
  github.com/coreos/etcd/etcdserver.(*authApplierV3).Apply()
      /home/gyuho/go/src/github.com/coreos/etcd/etcdserver/apply_auth.go:58 +0xe3
  github.com/coreos/etcd/etcdserver.(*EtcdServer).applyEntryNormal()
      /home/gyuho/go/src/github.com/coreos/etcd/etcdserver/server.go:1351 +0x420
  github.com/coreos/etcd/etcdserver.(*EtcdServer).apply()
      /home/gyuho/go/src/github.com/coreos/etcd/etcdserver/server.go:1274 +0x574
  github.com/coreos/etcd/etcdserver.(*EtcdServer).applyEntries()
      /home/gyuho/go/src/github.com/coreos/etcd/etcdserver/server.go:913 +0x189
  github.com/coreos/etcd/etcdserver.(*EtcdServer).applyAll()
      /home/gyuho/go/src/github.com/coreos/etcd/etcdserver/server.go:775 +0xc3
  github.com/coreos/etcd/etcdserver.(*EtcdServer).run.func6()
      /home/gyuho/go/src/github.com/coreos/etcd/etcdserver/server.go:737 +0x53
  github.com/coreos/etcd/pkg/schedule.(*fifo).run()
      /home/gyuho/go/src/github.com/coreos/etcd/pkg/schedule/schedule.go:158 +0x10a

Previous write at 0x00c422388ea0 by goroutine 38:
  runtime.closechan()
      /usr/local/go/src/runtime/chan.go:313 +0x0
  github.com/coreos/etcd/mvcc.(*watchStream).Close()
      /home/gyuho/go/src/github.com/coreos/etcd/mvcc/watcher.go:153 +0x17b
  github.com/coreos/etcd/etcdserver/api/v3rpc.(*serverWatchStream).close()
      /home/gyuho/go/src/github.com/coreos/etcd/etcdserver/api/v3rpc/watch.go:394 +0x57
  github.com/coreos/etcd/etcdserver/api/v3rpc.(*watchServer).Watch()
      /home/gyuho/go/src/github.com/coreos/etcd/etcdserver/api/v3rpc/watch.go:157 +0x533
  github.com/coreos/etcd/etcdserver/etcdserverpb._Watch_Watch_Handler()
      /home/gyuho/go/src/github.com/coreos/etcd/etcdserver/etcdserverpb/rpc.pb.go:3480 +0xd5
  github.com/grpc-ecosystem/go-grpc-prometheus.StreamServerInterceptor()
      /home/gyuho/go/src/github.com/grpc-ecosystem/go-grpc-prometheus/server.go:40 +0x1ce
  github.com/coreos/etcd/etcdserver/api/v3rpc.newStreamInterceptor.func1()
      /home/gyuho/go/src/github.com/coreos/etcd/etcdserver/api/v3rpc/interceptor.go:93 +0x135
  google.golang.org/grpc.(*Server).processStreamingRPC()
      /home/gyuho/go/src/google.golang.org/grpc/server.go:898 +0x685
  google.golang.org/grpc.(*Server).handleStream()
      /home/gyuho/go/src/google.golang.org/grpc/server.go:985 +0x143a
  google.golang.org/grpc.(*Server).serveStreams.func1.1()
      /home/gyuho/go/src/google.golang.org/grpc/server.go:551 +0xac

Goroutine 55 (running) created at:
  github.com/coreos/etcd/pkg/schedule.NewFIFOScheduler()
      /home/gyuho/go/src/github.com/coreos/etcd/pkg/schedule/schedule.go:71 +0x30d
  github.com/coreos/etcd/etcdserver.(*EtcdServer).run()
      /home/gyuho/go/src/github.com/coreos/etcd/etcdserver/server.go:630 +0x1ca

Goroutine 38 (running) created at:
  google.golang.org/grpc.(*Server).serveStreams.func1()
      /home/gyuho/go/src/google.golang.org/grpc/server.go:549 +0xb8
  google.golang.org/grpc/transport.(*http2Server).operateHeaders()
      /home/gyuho/go/src/google.golang.org/grpc/transport/http2_server.go:333 +0x1493
  google.golang.org/grpc/transport.(*http2Server).HandleStreams()
      /home/gyuho/go/src/google.golang.org/grpc/transport/http2_server.go:400 +0xaad
  google.golang.org/grpc.(*Server).serveStreams()
      /home/gyuho/go/src/google.golang.org/grpc/server.go:547 +0x1d1
  google.golang.org/grpc.(*Server).serveHTTP2Transport()
      /home/gyuho/go/src/google.golang.org/grpc/server.go:540 +0x68c
  google.golang.org/grpc.(*Server).handleRawConn()
      /home/gyuho/go/src/google.golang.org/grpc/server.go:507 +0x5c6
==================
panic: send on closed channel

goroutine 20886 [running]:
github.com/coreos/etcd/mvcc.(*watcher).send(0xc422849f80, 0x18, 0xc423f1ebe0, 0x1, 0x1, 0x54, 0x0, 0x4387ee)
	/home/gyuho/go/src/github.com/coreos/etcd/mvcc/watchable_store.go:518 +0x4ad
github.com/coreos/etcd/mvcc.(*watchableStore).notify(0xc420244280, 0x54, 0xc423f1ebc0, 0x1, 0x1)
	/home/gyuho/go/src/github.com/coreos/etcd/mvcc/watchable_store.go:429 +0x22d
github.com/coreos/etcd/mvcc.(*watchableStoreTxnWrite).End(0xc423f1eb00)
	/home/gyuho/go/src/github.com/coreos/etcd/mvcc/watchable_store_txn.go:43 +0x385
github.com/coreos/etcd/etcdserver.(*applierV3backend).Txn(0xc4255d4340, 0xc42284d0e0, 0xc42284d0e0, 0x1, 0x28)
	/home/gyuho/go/src/github.com/coreos/etcd/etcdserver/apply.go:361 +0x361
github.com/coreos/etcd/etcdserver.(*quotaApplierV3).Txn(0xc4255d4380, 0xc42284d0e0, 0xc4260ae088, 0xc42284d0e0, 0x0)
	/home/gyuho/go/src/github.com/coreos/etcd/etcdserver/apply.go:784 +0xab
github.com/coreos/etcd/etcdserver.(*authApplierV3).Txn(0xc4260ae050, 0xc42284d0e0, 0xc423f36db0, 0x0, 0x0)
	/home/gyuho/go/src/github.com/coreos/etcd/etcdserver/apply_auth.go:169 +0xee
github.com/coreos/etcd/etcdserver.(*applierV3backend).Apply(0xc4255d4340, 0xc423122000, 0xf0cf2a)
	/home/gyuho/go/src/github.com/coreos/etcd/etcdserver/apply.go:119 +0x625
github.com/coreos/etcd/etcdserver.(*quotaApplierV3).Apply(0xc4255d4380, 0xc423122000, 0xc4260ae080)
	<autogenerated>:1 +0x62
github.com/coreos/etcd/etcdserver.(*authApplierV3).Apply(0xc4260ae050, 0xc423122000, 0x0)
	/home/gyuho/go/src/github.com/coreos/etcd/etcdserver/apply_auth.go:58 +0xe4
github.com/coreos/etcd/etcdserver.(*EtcdServer).applyEntryNormal(0xc42133e000, 0xc42314b5d8)
	/home/gyuho/go/src/github.com/coreos/etcd/etcdserver/server.go:1351 +0x421
github.com/coreos/etcd/etcdserver.(*EtcdServer).apply(0xc42133e000, 0xc42255f560, 0x2, 0x2, 0xc423174690, 0xc42314b688, 0x478a48, 0x66f51a4cecd)
	/home/gyuho/go/src/github.com/coreos/etcd/etcdserver/server.go:1274 +0x575
github.com/coreos/etcd/etcdserver.(*EtcdServer).applyEntries(0xc42133e000, 0xc423174690, 0xc4222389a0)
	/home/gyuho/go/src/github.com/coreos/etcd/etcdserver/server.go:913 +0x18a
github.com/coreos/etcd/etcdserver.(*EtcdServer).applyAll(0xc42133e000, 0xc423174690, 0xc4222389a0)
	/home/gyuho/go/src/github.com/coreos/etcd/etcdserver/server.go:775 +0xc4
github.com/coreos/etcd/etcdserver.(*EtcdServer).run.func6(0x7f378a6622b0, 0xc423000540)
	/home/gyuho/go/src/github.com/coreos/etcd/etcdserver/server.go:737 +0x54
github.com/coreos/etcd/pkg/schedule.(*fifo).run(0xc4231168a0)
	/home/gyuho/go/src/github.com/coreos/etcd/pkg/schedule/schedule.go:158 +0x10b
created by github.com/coreos/etcd/pkg/schedule.NewFIFOScheduler
	/home/gyuho/go/src/github.com/coreos/etcd/pkg/schedule/schedule.go:71 +0x30e
exit status 2
FAIL	github.com/coreos/etcd/clientv3/integration	58.823s
```

--- a/mvcc/watchable_store.go
+++ b/mvcc/watchable_store.go
@@ -144,13 +144,15 @@ func (s *watchableStore) watch(key, end []byte, startRev int64, id WatchID, ch c
 func (s *watchableStore) cancelWatcher(wa *watcher) {
 	for {
 		s.mu.Lock()
-
 		if s.unsynced.delete(wa) {
 			slowWatcherGauge.Dec()
 			break
 		} else if s.synced.delete(wa) {
 			break
 		} else if wa.compacted {
+			break
+		} else if wa.ch == nil {
+			// already canceled (e.g., cancel/close race)
 			break
 		}
 
@@ -177,6 +179,7 @@ func (s *watchableStore) cancelWatcher(wa *watcher) {
 	}
 
 	watcherGauge.Dec()
+	wa.ch = nil
 	s.mu.Unlock()
 }
 
@@ -425,7 +428,6 @@ func (s *watchableStore) notify(rev int64, evs []mvccpb.Event) {
 		if eb.revs != 1 {
 			plog.Panicf("unexpected multiple revisions in notification")
 		}
-
 		if w.send(WatchResponse{WatchID: w.id, Events: eb.evs, Revision: rev}) {
 			pendingEventsGauge.Add(float64(len(eb.evs)))
 		} else {

--- a/mvcc/watchable_store_test.go
+++ b/mvcc/watchable_store_test.go
@@ -539,3 +539,49 @@ func TestWatchVictims(t *testing.T) {
 	default:
 	}
 }
+
+// TestStressWatchCancelClose tests closing a watch stream while
+// canceling its watches.
+func TestStressWatchCancelClose(t *testing.T) {
+	b, tmpPath := backend.NewDefaultTmpBackend()
+	s := newWatchableStore(b, &lease.FakeLessor{}, nil)
+
+	defer func() {
+		s.store.Close()
+		os.Remove(tmpPath)
+	}()
+
+	testKey, testValue := []byte("foo"), []byte("bar")
+	var wg sync.WaitGroup
+	readyc := make(chan struct{})
+	wg.Add(100)
+	for i := 0; i < 100; i++ {
+		go func() {
+			defer wg.Done()
+			w := s.NewWatchStream()
+			ids := make([]WatchID, 10)
+			for i := range ids {
+				ids[i] = w.Watch(testKey, nil, 0)
+			}
+			<-readyc
+			wg.Add(1 + len(ids)/2)
+			for i := range ids[:len(ids)/2] {
+				go func(n int) {
+					defer wg.Done()
+					w.Cancel(ids[n])
+				}(i)
+			}
+			go func() {
+				defer wg.Done()
+				w.Close()
+			}()
+		}()
+	}
+
+	close(readyc)
+	for i := 0; i < 100; i++ {
+		s.Put(testKey, testValue, lease.NoLease)
+	}
+
+	wg.Wait()
+}

--- a/mvcc/watcher.go
+++ b/mvcc/watcher.go
@@ -129,16 +129,25 @@ func (ws *watchStream) Chan() <-chan WatchResponse {
 func (ws *watchStream) Cancel(id WatchID) error {
 	ws.mu.Lock()
 	cancel, ok := ws.cancels[id]
+	w := ws.watchers[id]
 	ok = ok && !ws.closed
-	if ok {
-		delete(ws.cancels, id)
-		delete(ws.watchers, id)
-	}
 	ws.mu.Unlock()
+
 	if !ok {
 		return ErrWatcherNotExist
 	}
 	cancel()
+
+	ws.mu.Lock()
+	// The watch isn't removed until cancel so that if Close() is called,
+	// it will wait for the cancel. Otherwise, Close() could close the
+	// watch channel while the store is still posting events.
+	if ww := ws.watchers[id]; ww == w {
+		delete(ws.cancels, id)
+		delete(ws.watchers, id)
+	}
+	ws.mu.Unlock()
+
 	return nil
 }
 

